homeassistant/components/motion_blinds/cover.py

```diff
@@ -224,6 +234,27 @@ def is_closed(self):
             return None
         return self._blind.position == 100
 
+    @property
+    def is_opening(self):
+        """Return if the cover is opening or not."""
+        if self.is_closed is None:
+            return None
+        return self._moving == State.Opening
+
+    @property
+    def is_closing(self):
+        """Return if the cover is closing or not."""
+        if self.is_closed is None:
+            return None
+        return self._moving == State.Closing
+
+    def _set_moving_state(self, new_position, old_position):
+        """Set the moving status for a position move."""
+        if new_position > old_position:
```

Comment:
We don't allow device and service interfacing integrations to calculate state. None of the added states here seem to be gotten from api information. I suggest we revert this PR.

Reply:
> We don't allow device and service interfacing integrations to calculate state. None of the added states here seem to be gotten from api information.

I agree. Sorry I missed that, I was a bit too focused on the code review vs the standards on this one.

> I suggest we revert this PR.

@starkillerOG   The implementation needs to get the opening and closing states from the api.  Do you want to do the revert?

Reply:
Well the command that issues the move does give a acknowledgement message that the device actually started moving, if this ACK message is not received or contains inproper values a exception will be raised and the state would not be set to moving. So in a sence the API does provide conformation that the blind actually started moving.

Regarding when it finishes moving, the position of the blind is retrieved from the blind itself here: https://github.com/home-assistant/core/blob/23dd3ee953aefd77baa86ea81caab35c63406cf7/homeassistant/components/motion_blinds/cover.py#L276
The not_moving state is gotton from the API by seeing that the retrieved position does not change anymore (definition of not moving).

So I would argue that the moving states are therefore not assumed but retrieved from the device.

Also this was a feature request from a user: https://community.home-assistant.io/t/motion-blinds-opening-closing-state/406417

Reply:
However if you want me to revert this PR, I will, just wanted to put in my opinion :)
Please confirm how to proceed.

Reply:
The library doesn't raise if the there's a missing ack:
https://github.com/starkillerOG/motion-blinds/blob/3027c2569b96b364217febb757b5f2b265386da4/motionblinds/motion_blinds.py#L947-L954

Am I looking at the wrong place?

The response parsing also doesn't seem to consistently raise exceptions if the response parsing encounters missing or incorrect data. There's a variation of logging and raising a `ParseException`.

So currently we can't trust these things to mean accurate state information.

If the api has a contract from its request responses to describe the device movement state we can use that as state information, yes. But currently the library doesn't seem to forward this information consistently to the library user.

It would be better if the library kept track of the movement state information so that each library user doesn't need to do this.

Reply:
the raise is on line 481 inside the _send function:

Reply:
So for instance the `blind.Open` function calls `blind._write` which calls `blind._gateway._write_subdevice` which calls `gateway._send` and that has the raise (of the socket.timeout after 3x sending the message) on line 481.

Reply:
It is true that the parsing of the responses catches a lot of errors, since there are quite some models that provide diffrent amounts of data. Some older models are missing some of the attributes. Newer models with the latest firmware installed on the blind itself schould have all the values in every response. (unfortunately the firmware of the blind itself can not be updated).

Reply:
The problem with moving this logic to the library is that the API does not provide a push event when the blind stops moving. Therefore during movement the blind needs to be polled at a high rate (every 5 seconds) to see the postion changing untill the position stops changing in order to determe the blind has stopped moving.

Having the polling included in the library would mean that a .Open function would block for up to possibly 2 minutes for very large blinds (like outdoor awnings) that take quite some time to fully open/close. Since the library is not yet async I don't think that is a good idea. (with the async lock that would also block the .Stop function for instance).

Reply:
I will make a PR to revert the canges tonight.

Reply:
Here is the PR to revert the changes: https://github.com/home-assistant/core/pull/69299
Still find it a pitty, but I do understand.